### DISCUSSION TITLE
FileManager example for custom folder

### DISCUSSION THREAD

**Discussion by markdav-is** [oqtane/oqtane.framework#1700] [Q&A]
I'm excited about the FileManager control and spent a few hours yesterday looking at how it's used in the admin modules as well as looking at the control source code.  My scenario is I'd like to specify a custom folder called "uploads" and when a file is uploaded, I'd like to pass that fileId to a controller to process the file.   The issue I'm having is that the FileID from the FileManager is always 0 (zero).

I feel like I'm missing a step here.  The filemanager reports a successful upload, but it's not anywhere if I go into the file manager admin module.

```
<FileManager @ref="_filemanager"
                            UploadMultiple="false" Folder="Uploads"
                            ShowFolders="false" ShowFiles="false" Filter="xlsx" />
<button type="button" class="btn btn-success" @onclick="Import">@Localizer["Import"]</button>

 private async Task Import() {
        lookupId = _filemanager.GetFileId();  << ID is always 0 (zero)
        await ImportService.ImportLookupAsync(lookupId, ModuleState.ModuleId);
    }
```

**Comment by leigh-pointer** [comment]
You are missing a ref to the lookupId in the FileManager component.  

**Try this**

```
<FileManager @ref="_filemanager" FileId="@lookupId"
                            UploadMultiple="false" Folder="Uploads"
                            ShowFolders="false" ShowFiles="false" Filter="xlsx" />
<button type="button" class="btn btn-success" @onclick="Import">@Localizer["Import"]</button>

@code {

int lookupId = -1;

 private async Task Import() {
        lookupId = _filemanager.GetFileId();  
        await ImportService.ImportLookupAsync(lookupId, ModuleState.ModuleId);
    }
}
```

**Reply by markdav-is** [reply]
The file ID is always 0 (zero) for me.  I ended up making a custom version of FileMakager that has a callback for the file being uploaded.  New files have no Ids, so I must look the file up my name to get a handle to it so I can pass the ID.  It's a little wonky, but it seems to be working.

**Comment by sbwalker** [comment]
@markdav-is does the "uploads" folder actually exist in your site? Does it have the correct permissions? The FileManager component is used extensively throughout Oqtane to allow users to upload or select files based on the folders for which they have permissions. In general, the recommendation is to not hardcode folder names but use FolderIds. If you absolutely need to store files in a specify folder then use the folder service to lookup by the folder by path and then pass the FolderId to the FileManager component.

**Comment by sbwalker** [comment]
I took a closer look at the FileManager and the Folder parameter has a comment "// optional - for setting a specific folder by default ( only relevant for host functions )". The original purpose of this parameter was that there needed to be a way to use the FileManager for uploading nuget packages to specific host level folders for installation. However, the framework has since been refactored so that all nuget packages are uploaded to a single "Packages" folder. So it may make more sense now if the Folder parameter allowed a developer to pass a folder path ( which would then be converted to a FolderId internally by doing a lookup ).

**Reply by sbwalker** [reply]
this is resolved now in PR #1709
